### ISSUE TITLE
routed_decoder slows down pipeline if placed after shuffle

### ISSUE THREAD

**Issue by MatthewCaseres** [pytorch/data#668]
### 🐛 Describe the bug

I ran the following code - 
```py
from torchdata.datapipes.iter import FileLister
from torch.utils.data.datapipes.utils.decoder import imagehandler
import time
    
# Decode before shuffling
dp1 = FileLister("./data") \
    .open_files(mode="b") \
    .load_from_tar() \
    .routed_decode(imagehandler("torch")) \
    .shuffle(buffer_size=100) 
t1 = time.time()
for i, (x, y) in enumerate(dp1):
    if i == 100:
        break
t2 = time.time()
print("dif is ", t2 - t1)
# Decode after shuffling
dp2 = FileLister("./data") \
    .open_files(mode="b") \
    .load_from_tar() \
    .shuffle(buffer_size=100) \
    .routed_decode(imagehandler("torch")) 

t3 = time.time()
for i, (x, y) in enumerate(dp2):
    if i == 100:
        break
t4 = time.time()
print("dif is ", t4 - t3)
```
The second datapipe is much slower
```
dif is  5.465105056762695
dif is  146.25729250907898
```

I don't see this documented anywhere. This causes problems with sharding, because `It is crucial to add ShardingFilter after Shuffler to ensure that all worker processes have the same order of data for sharding.` but also `Place ShardingFilter (datapipe.sharding_filter) as early as possible in the pipeline, especially before expensive operations such as decoding`.

So the ordering has to be shuffle -> shard -> decode. But this seems like it will cause problems because placing shuffle before decode seems to cause performance problems, as discussed above.

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.4 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: version 3.22.3
Libc version: glibc-2.31

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 16:56:21)  [GCC 10.3.0] (64-bit runtime)
Python platform: Linux-5.13.0-1031-aws-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: 
GPU 0: Tesla T4
GPU 1: Tesla T4
GPU 2: Tesla T4
GPU 3: Tesla T4

Nvidia driver version: 510.73.08
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.22.4
[pip3] torch==1.12.0
[pip3] torch-model-archiver==0.5.3b20220226
[pip3] torch-workflow-archiver==0.2.4b20220513
[pip3] torchaudio==0.12.0
[pip3] torchdata==0.4.0
[pip3] torchserve==0.6.0b20220513
[pip3] torchtext==0.13.0
[pip3] torchvision==0.13.0
[conda] blas                      2.115                       mkl    conda-forge
[conda] blas-devel                3.9.0            15_linux64_mkl    conda-forge
[conda] captum                    0.5.0                         0    pytorch
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] libblas                   3.9.0            15_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            15_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            15_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            15_linux64_mkl    conda-forge
[conda] magma-cuda116             2.6.1                         0    pytorch
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.22.4                   pypi_0    pypi
[conda] pytorch                   1.12.0          py3.9_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-model-archiver      0.5.3                    py39_0    pytorch
[conda] torch-workflow-archiver   0.2.4                    py39_0    pytorch
[conda] torchaudio                0.12.0               py39_cu116    pytorch
[conda] torchserve                0.6.0                    py39_0    pytorch
[conda] torchtext                 0.13.0                     py39    pytorch
[conda] torchvision               0.13.0               py39_cu116    pytorch
```

**Comment by NivekT** [comment]
Hi, thanks for opening this issue.

1. Just as a sanity check, if you swap the order of execution in your script, does the resulting times change?
2. What are the source files? We would like to reproduce the issue on our end if possible.
3. Are you observing this behavior for other DataPipes? Note that we plan to deprecate `routed_decode` in either the next release or the one after.

**Comment by MatthewCaseres** [comment]
The source files are some jpeg images, each tar file has like 500 images. I noticed the time it took to load images was greatly influenced by buffer size, I don't think this is to do with the order of the statements run. I will provide a reproduction tomorrow.

I assumed `routed_decode` was transitioning from main PyTorch repo over here, if not, what is the right pattern for  decoding?

**Comment by MatthewCaseres** [comment]
https://github.com/MatthewCaseres/torchdata-slowpipe-repro

Results from running script - 
```
decode before shuffle, buffer size 10
/opt/conda/envs/pytorch/lib/python3.9/site-packages/torch/utils/data/datapipes/utils/common.py:173: FutureWarning: `RoutedDecoderIterDataPipe()` and its functional API `.routed_decode()` are deprecated since 1.12 and will be removed in 1.13.
See https://github.com/pytorch/data/issues/163 for details.
  warnings.warn(msg, FutureWarning)
100th image: 0.32980823516845703
decode after shuffle, buffer size 10
100th image: 1.446854829788208
decode before shuffle, buffer size 100
100th image: 0.30071139335632324
decode after shuffle, buffer size 100
100th image: 2.812696933746338
decode before shuffle, buffer size 1000
100th image: 0.1922600269317627
decode after shuffle, buffer size 1000
100th image: 17.595024585723877
```

**Comment by NivekT** [comment]
Thanks for the update. What is the difference between the sizes (in bytes) of the buffer depending in whether you decode before or after? Maybe call `sys. getsizeof()` on `Shuffler._buffer`?

**Comment by ejguan** [comment]
Are those archives `tar` files or `tar.gz` files?

**Comment by NivekT** [comment]
> Are those archives `tar` files or `tar.gz` files?

They are `tar.gz` files as shown in the README (https://github.com/MatthewCaseres/torchdata-slowpipe-repro).

**Comment by ejguan** [comment]
It might the problem that decoding after shuffle requires changing the position of file handles of `tar` archives (deserializing + seeking), which taking more time than reading sequentially.

This might be resolved by de-compressing the archive first. It's worth experimenting on the dataset.

**Comment by MatthewCaseres** [comment]
Issue is resolved when tarfile is just .tar and not .tar.gzip. I had to decompress and recompress using -cvf flags then it was good. This can be closed as this is the intended usage pattern and why we have the decompressors?

**Comment by ejguan** [comment]
You should be able to decompress via `gzip -d xxx.tar.gz` the you don't have to re-create tar archive.

**Comment by NivekT** [comment]
I opened a PR #688 so that people are aware that shuffling early can impact the performance of operations that benefit from sequential read.
